pkgs/development/interpreters/python/fetch-python-requirements.nix

```diff
@@ -0,0 +1,127 @@
+# api responses from pypi.org to only contain files for which the
+# release date is lower than the specified maxDate.
+
+{
+  # this specifies the python version for which the packages should be downloaded
+  python,
+}:
+
+{
+  lib
+, cacert
+, curl
+, stdenv
+, buildPackages
+, python3
+, ...
+}:
+let
+
+  # we use mitmproxy to filter the pypi responses
+  pythonWithMitmproxy = python3.withPackages (ps: [ ps.mitmproxy ps.python-dateutil ]);
+
+  fetchPythonRequirements = {
+    maxDate,        # maximum release date for packages
+    outputHash,     # hash for the fixed output derivation
+    requirements,   # list of strings of specs
+
+    pipFlags ? [],
+  }:
+    let
+      pythonAttrName = lib.stringAsChars (x: if x == "." then "" else x) python.libPrefix;
+
+      # We need to execute pip from the python version for which we want to download packages.
+      # Pip accepts '--python-version', but this works only for wheel packages.
+      pythonPip = buildPackages."${pythonAttrName}".withPackages (ps: [
+        ps.pip
+        ps.setuptools
+        ps.pkgconfig
+      ]);
+
+      # Normalize package names:
+      #   "Example_Package[extra]" -> "example-package"
+      normalizedNames = map (req:
+        lib.toLower (lib.replaceStrings [ "_" ] [ "-" ] (lib.elemAt (lib.splitString "[" req) 0))
+      ) requirements;
+
+      # fixed output derivation containing packages,
+      # each being symlinked from it's normalized name
+      # Example:
+      #   "$out/werkzeug" will point to "$out/Werkzeug-0.14.1-py2.py3-none-any.whl"
+      self = stdenv.mkDerivation {
+
+        name = "multiple-pypi-packages";
+
+        outputHashMode = "recursive";
+        outputHashAlgo = "sha256";
+        inherit outputHash;
+
+        nativeBuildInputs = with python3.pkgs; [ pythonWithMitmproxy pythonPip curl cacert ];
+
+        phases = [ "buildPhase" ];
+
+        buildPhase = ''
+          # the script.py will read this date
+          export MAX_DATE=${maxDate}
+          pretty=$(python -c 'import os; import dateutil.parser; print(dateutil.parser.parse(os.getenv("MAX_DATE")))')
+          echo "selected maximum release date for python packages: $pretty"
+
+          # find free port for proxy
+          proxyPort=$(python -c '\
+          import socket
+          s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+          s.bind(("", 0))
+          print(s.getsockname()[1])
+          s.close()')
+
+          # start proxy to filter pypi responses
+          # mitmproxy wants HOME set
+          # mitmdump == mitmproxy without GUI
+          HOME=$(pwd) mitmdump \
+            --listen-port "$proxyPort" \
+            --ignore-hosts '.*files.pythonhosted.org.*'\
+            --script ${./filter-pypi-responses.py} &
+
+          # wait for proxy to come up
+          while sleep 0.5; do
+            timeout 5 curl --proxy http://localhost:$proxyPort http://pypi.org 2>/dev/null && break
```

Comment:
Add curl --fail flag? (Maybe it's not stricktly needed, but feels like a good idea to me.)

Reply:
As I understand it, --fail just removes http error messages from the output. But since we have 2>/dev/null, it wouldn't make any difference at all, or would it? Also, the errors we are receiving here are `connection refused` errors which are not affected by this setting anyways.

Reply:
The `--fail` flag also makes curl exit with an error (22) instead of success (0) if there is an e.g. 404. So it's generally a good thing to use in scripts.

> Also, the errors we are receiving here are connection refused errors which are not affected by this setting anyways.

Ok. I'm just always on the alert when I see curl commands without `--fail`. I would personally add it, for peace of mind :smile:

Reply:
Oh, I wasn't aware of the error codes. Fixed in 371cd6bffecb0e64cfece6438bfc13d47d63fe7f